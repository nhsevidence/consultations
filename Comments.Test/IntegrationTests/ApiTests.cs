using Comments.Test.Infrastructure;
using Comments.ViewModels;
using Newtonsoft.Json;
using Shouldly;
using Shouldly.ShouldlyExtensionMethods;
using System;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Text;
using System.Threading.Tasks;
using Xunit;

namespace Comments.Test.IntegrationTests
{
    public class ApiTests : TestBase
    {
        [Fact]
        public async Task GetDocumentReturnsEmptyFeedIII()
        {
            //Arrange (in the base constructor for this one.)

            // Act
            var response = await _client.GetAsync("/api/Document?consultationId=1&documentId=1&chapterSlug=some-chapter");
            response.EnsureSuccessStatusCode();
            var responseString = await response.Content.ReadAsStringAsync();

            // Assert
            responseString.ShouldMatchApproved();
        }

        [Theory]
        [InlineData(1, 1, null)]
        [InlineData(1, 1, 1)]
        [InlineData(int.MaxValue, int.MaxValue, null)]
        [InlineData(int.MaxValue, int.MaxValue, 1)]
        public async Task CreateComment(int locationId, int consultationId, int? documentId)
        {
            //Arrange
            var comment = new ViewModels.Comment(locationId, consultationId, documentId, "chapter-slug", null, null, null, null, null, null, 0, DateTime.Now, Guid.Empty, "comment text");
            var content = new StringContent(JsonConvert.SerializeObject(comment), Encoding.UTF8, "application/json");

            // Act
            var response = await _client.PostAsync("api/Comment", content);
            response.EnsureSuccessStatusCode();
            var responseString = await response.Content.ReadAsStringAsync();

            // Assert
            response.StatusCode.ShouldBe(HttpStatusCode.Created);
            var deserialisedComment = JsonConvert.DeserializeObject<ViewModels.Comment>(responseString);
            deserialisedComment.CommentId.ShouldBeGreaterThan(0);
        }

        [Fact]
        public async Task Insert_Multiple_Comments_And_Read_Them()
        {
            //Arrange
            ResetDatabase();
            const int locationId = 1;
            const int consultationId = 1;
            const int documentId = 1;
            await CreateComment(locationId, consultationId, documentId);
            await CreateComment(locationId, consultationId, documentId); //duplicate comment. totally valid.
            await CreateComment(2, consultationId, documentId); //different location id, this should be in the result set
            await CreateComment(locationId, 2, documentId); //different consultation id, this shouldn't be in the result set
            await CreateComment(locationId, consultationId, 2); //different document id, this shouldn't be in the result set

            // Act
            var response = await _client.GetAsync($"/api/Document?consultationId={consultationId}&documentId={documentId}");
            response.EnsureSuccessStatusCode();
            var responseString = await response.Content.ReadAsStringAsync();

            // Assert
            var deserialisedResponse = JsonConvert.DeserializeObject<DocumentViewModel>(responseString);
            deserialisedResponse.Comments.Count().ShouldBe(3);
            responseString.ShouldMatchApproved();
        }

        [Fact]
        public async Task GetDocumentReturnsPopulatedFeed()
        {
            // Arrange
            ResetDatabase();
            var consultationId = 1;
            var documentId = 1;
            var commentText = Guid.NewGuid().ToString();
            var questionText = Guid.NewGuid().ToString();
            var answerText = Guid.NewGuid().ToString();

            AddCommentsAndQuestionsAndAnswers(consultationId, documentId, commentText, questionText, answerText);
            
            // Act
            var response = await _client.GetAsync($"/api/Document?consultationId={consultationId}&documentId={documentId}");
            response.EnsureSuccessStatusCode();
            var responseString = await response.Content.ReadAsStringAsync();

            // Assert
            responseString.ShouldMatchApproved();
            //todo: get the below all working:

            //var deserialisedResponse = JsonConvert.DeserializeObject<DocumentViewModel>(responseString);
            //deserialisedResponse.Comments.Single().CommentText.ShouldBe(commentText);
            //deserialisedResponse.Questions.Single().QuestionText.ShouldBe(questionText);
            //deserialisedResponse.Questions.Single().Answers.Single().AnswerText.ShouldBe(answerText);
        }
    }
}
