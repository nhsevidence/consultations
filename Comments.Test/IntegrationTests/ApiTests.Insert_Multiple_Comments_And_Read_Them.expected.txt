{
  "consultation": {
    "documents": [
      {
        "documentId": 2,
        "supportsComments": false,
        "isSupportingDocument": false,
        "title": null,
        "chapters": null
      },
      {
        "documentId": 2,
        "supportsComments": true,
        "isSupportingDocument": false,
        "title": "Unstable angina and NSTEMI",
        "chapters": [
          {
            "slug": "introduction",
            "title": "Introduction"
          },
          {
            "slug": "patient-centred-care",
            "title": "Patient-centred care"
          },
          {
            "slug": "key-priorities-for-implementation",
            "title": "Key priorities for implementation"
          },
          {
            "slug": "guidance",
            "title": "1 Guidance"
          },
          {
            "slug": "notes-on-the-scope-of-the-guidance",
            "title": "2 Notes on the scope of the guidance"
          },
          {
            "slug": "implementation",
            "title": "3 Implementation"
          },
          {
            "slug": "research-recommendations",
            "title": "4 Research recommendations"
          },
          {
            "slug": "other-versions-of-this-guideline",
            "title": "5 Other versions of this guideline"
          },
          {
            "slug": "related-nice-guidance",
            "title": "6 Related NICE guidance"
          },
          {
            "slug": "updating-the-guideline",
            "title": "7 Updating the guideline"
          },
          {
            "slug": "appendix-a-the-guideline-development-group-and-nice-project-team",
            "title": "Appendix A: The Guideline Development Group and NICE project team"
          },
          {
            "slug": "appendix-b-the-guideline-review-panel",
            "title": "Appendix B: The Guideline Review Panel"
          },
          {
            "slug": "appendix-c-the-algorithm",
            "title": "Appendix C: The algorithm"
          },
          {
            "slug": "changes-after-publication",
            "title": "Changes after publication"
          },
          {
            "slug": "about-this-guideline",
            "title": "About this guideline"
          }
        ]
      },
      {
        "documentId": 2,
        "supportsComments": false,
        "isSupportingDocument": false,
        "title": null,
        "chapters": null
      },
      {
        "documentId": 2,
        "supportsComments": true,
        "isSupportingDocument": false,
        "title": "Sinusitis (acute): antimicrobial prescribing",
        "chapters": [
          {
            "slug": "overview",
            "title": "Overview"
          },
          {
            "slug": "background",
            "title": "Background"
          },
          {
            "slug": "recommendations",
            "title": "Recommendations"
          },
          {
            "slug": "symptoms-and-signs",
            "title": "Symptoms and signs"
          },
          {
            "slug": "summary-of-the-evidence",
            "title": "Summary of the evidence"
          },
          {
            "slug": "other-considerations",
            "title": "Other considerations"
          }
        ]
      }
    ],
    "reference": "GID-TA10232",
    "title": "For consultation comments",
    "consultationName": "Appraisal consultation",
    "startDate": "2018-04-01T00:00:00",
    "endDate": "2018-05-31T17:00:00",
    "consultationType": "Appraisal consultation",
    "resourceTitleId": null,
    "projectType": "TA",
    "productTypeName": "Technology appraisal guidance",
    "developedAs": null,
    "relevantTo": null,
    "consultationId": 1,
    "process": "TAG",
    "allowConsultationComments": true,
    "partiallyUpdatedProjectReference": null,
    "origProjectReference": null
  },
  "selectedConsultationDocumentId": 1,
  "chapterHTML": {
    "content": null,
    "sections": null,
    "slug": null,
    "title": "For consultation comments"
  },
  "comments": [
    {
      "commentId": 1,
      "lastModifiedDate": "0001-01-01T00:00:00",
      "lastModifiedByUserId": "00000000-0000-0000-0000-000000000000",
      "commentText": "comment text",
      "locationId": 1,
      "consultationId": 1,
      "documentId": 1,
      "chapterSlug": "chapter-slug",
      "sectionSlug": null,
      "rangeStart": null,
      "rangeStartOffset": null,
      "rangeEnd": null,
      "rangeEndOffset": null,
      "quote": null
    },
    {
      "commentId": 2,
      "lastModifiedDate": "0001-01-01T00:00:00",
      "lastModifiedByUserId": "00000000-0000-0000-0000-000000000000",
      "commentText": "comment text",
      "locationId": 2,
      "consultationId": 1,
      "documentId": 1,
      "chapterSlug": "chapter-slug",
      "sectionSlug": null,
      "rangeStart": null,
      "rangeStartOffset": null,
      "rangeEnd": null,
      "rangeEndOffset": null,
      "quote": null
    },
    {
      "commentId": 3,
      "lastModifiedDate": "0001-01-01T00:00:00",
      "lastModifiedByUserId": "00000000-0000-0000-0000-000000000000",
      "commentText": "comment text",
      "locationId": 3,
      "consultationId": 1,
      "documentId": 1,
      "chapterSlug": "chapter-slug",
      "sectionSlug": null,
      "rangeStart": null,
      "rangeStartOffset": null,
      "rangeEnd": null,
      "rangeEndOffset": null,
      "quote": null
    }
  ],
  "questions": []
}